go/vt/vtgate/healthcheck.go

```diff
@@ -0,0 +1,279 @@
+package vtgate
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/golang/glog"
+	pbq "github.com/youtube/vitess/go/vt/proto/query"
+	pbt "github.com/youtube/vitess/go/vt/proto/topodata"
+	"github.com/youtube/vitess/go/vt/tabletserver/tabletconn"
+	"github.com/youtube/vitess/go/vt/topo"
+	"golang.org/x/net/context"
+)
+
+// AddrUpdate updates about an endpoint and whether it is up.
+type AddrUpdate struct {
+	EndPoint *pbt.EndPoint
+	IsUp     bool
+}
+
+// NewHealthCheck creates a new HealthCheck object.
+func NewHealthCheck(endPoints []*pbt.EndPoint, addrUpdate <-chan *AddrUpdate, connTimeout time.Duration, retryDelay time.Duration) *HealthCheck {
+	hc := &HealthCheck{
+		addrToConns: make(map[string]*healthCheckConn),
+		targetToEPs: make(map[string]map[string]map[pbt.TabletType][]*pbt.EndPoint),
+		connTimeout: connTimeout,
+		retryDelay:  retryDelay,
+	}
+	hc.initConns(endPoints)
+	go hc.onAddrUpdate(addrUpdate)
+	return hc
+}
+
+// HealthCheck performs health checking and notifies downstream components about any changes.
+type HealthCheck struct {
+	mu          sync.RWMutex
+	addrToConns map[string]*healthCheckConn                              // addrToConns maps from address to the healthCheckConn object.
+	targetToEPs map[string]map[string]map[pbt.TabletType][]*pbt.EndPoint // targetToEPs maps from keyspace/shard/tablettype to a list of endpoints.
+	connTimeout time.Duration
+	retryDelay  time.Duration
+}
+
+// healthCheckConn contains details about an endpoint.
+type healthCheckConn struct {
+	mu         sync.RWMutex
+	target     *pbq.Target
+	cancelFunc context.CancelFunc
+	conn       tabletconn.TabletConn
+	stats      *pbq.RealtimeStats
+}
+
+// initConns starts health checking for the initial set of endpoints.
+// It waits till all endpoints are tried to connect once.
```

Comment:
As this happens only during VTGate starts, I think it is okay to wait for connectTimeout. It is connecting to all endpoints during initialization currently.

The case that I would like to avoid is:
If VTGate does not wait for all connections ready and reports ready for query, it could happen that there is no available connections for a keyspace/shard/tablettype, and VTGate fails fast by returning error for the query. If VTGate already tried all endpoints once, the fast-fail is valid. But in this case, VTGate generates error unnecessarily.